tests/suite.py

```diff
@@ -0,0 +1,136 @@
+                    decoded["suite_root"]
+                )
+            )
+        else:
+            raise ValueError("suite_root missing from configuration file.")
+
+        if "timeout" in decoded:
+            self.timeout = float(decoded["timeout"])
+        else:
+            self.timeout = 1.0
+
+        if 'paths' in decoded:
+            self.paths = decoded["paths"]
+
+    def _make_paths_absolute(self):
+        for path in self.paths:
+            rel_path = self.paths[path]
+            abs_path = os.path.abspath(
+                os.path.join(self.suite_root, rel_path)
+            )
+            if not os.path.exists(abs_path):
+                raise ValueError(
+                    "Path replacement for {short}: {abspath} does not exist "
+                    "(was expanded from {rel})".format(
+                        short=path,
+                        abspath=abs_path,
+                        rel=rel_path)
+                )
+            self.paths[path] = abs_path
+
+    def discover(self, directory=None):
+        for root, dirs, files in os.walk(
+                self.suite_root if directory is None else directory):
+            cur_dir_tests = {}
+            for f in files:
+                name, ext = os.path.splitext(f)
+                if ext == '.yml' and name[-5:] == '_test':
+                    full_path = os.path.abspath(os.path.join(root, f))
+                    cur_dir_tests[name[:-5]] = case.TestCase(full_path)
+            if len(cur_dir_tests) > 0:
+                rel_path = os.path.relpath(root, start=self.suite_root)
+                self._testcases[rel_path] = cur_dir_tests
+
+    def run(self, verbosity=None):
+        failed_count = 0
+
+        for test_dir in self._testcases:
+            failed = []
+            print("Running:", test_dir)
+            cases = self._testcases[test_dir]
+            for case in cases:
+                res = cases[case].run(self.paths, self.timeout)
+                if res["success"]:
+                    print('.', end="")
+                else:
+                    print('F', end="")
+                    failed.append({"case": case, "res": res})
+            print()
+
+            for fail in failed:
+                res = fail["res"]
+                print("Testcase", fail["case"], "failed:")
+                for prop in ["stdout", "stderr", "retval"]:
+                    if res[prop + "_diff"] is not None:
+                        sys.stdout.writelines(res[prop + "_diff"])
+
+            failed_count += len(failed)
+        return failed_count
+
+    @property
+    def suite_root(self):
+        return self._suite_root
+
+    @suite_root.setter
+    def suite_root(self, new_root):
+        if not os.path.isdir(new_root):
+                raise ValueError("Invalid suite_root setting: {dir} is not a"
+                                 "directory.".format(dir=new_root))
+        self._suite_root = os.path.abspath(new_root)
+
+
+if __name__ == '__main__':
```

Comment:
All this main function is duplicated in the `runner.py` file. I am not very experienced with Python and I am not sure if this is intended or not, but I guess that if you want to have a main function in both files you can extract out the common code into a separated file.

Reply:
Nope, not required and not intended. Just a copy + paste fail, I simply forgot to remove the code here.